Adds a way to preserve user-preferred mesh grid state

mockoocy opened the pull request:
Fixes a bug / bad UX in which caused mesh grid to became unhidden after certain operations like DataCollection or Phase change.

The steps to reproduce the bugs were as following:
- mount a sample
- open `draw_grid` control
- draw a few grids
- make some of them hidden
- right click on canvas, press `Data Collection` button
- run the task
After the task was done, each of the grids was shown again.
It's also worth adding that in the demo beamline many of the motors seem to not be mocked, so the behavior can't be observed. 


The reason for this behavior was further described in this Pull request https://github.com/mxcube/mxcubecore/pull/1044 for mxcubecore. This PR will be marked as draft until the mxcubecore one is merged.

The fix writes to state set by user to `Shape.user_state` attribute on the mxcubecore side.
When the data collection or phase change is finished, then the user-preferred state is restored by setting `Shape.state` to be equal to `Shape.user_state`.

mockoocy commented:
Turns out the same issue happens for the zoom and focus controls. Will also try to fix it for these in this PR.

fabcor-maxiv commented:
> Turns out the same issue happens for the zoom and focus controls. Will also try to fix it for these in this PR.

@mockoocy Note that there was work on focus and zoom controls very recently: <https://github.com/mxcube/mxcubeweb/pull/1457>. I do not know if it is related or not.

mockoocy committed:
Added a way to ensure that user state persists

mockoocy committed:
added a way to indicate that the grid's been hidden by user

mockoocy committed:
adding a way to preserve Shape.state when it is recreated

mockoocy commented:
added some changes; 
now the UI has some indicator that specific grid is being hidden (it appears with a darker background)
I have also changed `HWR.beamline.sample_view.add_shape_from_mpos` function, also when updating shapes there are optionally more arguments passed to it, so that `shape` and `user_shape` are not changed when grids is being redrawn sometimes.

marcus-oscarsson commented:
Thanks @mockoocy, do you happen to have a screenshot of the changes ?

mockoocy commented:
> Thanks @mockoocy, do you happen to have a screenshot of the changes ?

![image](https://github.com/user-attachments/assets/d83e0b38-d3d7-4f8f-9476-d5b4681e0e6a)

fabcor-maxiv commented on ui/src/components/SampleView/SampleView.css:
```diff
@@ -176,3 +176,7 @@
 }
+
+.hid-by-user {
+  background-color: #aaa !important;
```
Isn't it usually bad practice to use `!important` in CSS?

mockoocy replied:
You're right. I switched to using data attributes to override bootstrap's default css.

elmjag commented:
@marcus-oscarsson I think we have different ideas how this mesh hidding/un-hidding is suppose to work.

Like what is the actual desired behavior. Perhaps we can discuss this in a video meeting? I'm at least a bit confused right now.

mockoocy committed:
Changes to GridForm styling

using data attributes to override bootstrap css

added tooltip for Show/Hide button

changed color for hidden grids

make hidden grid's text italic

marcus-oscarsson commented:
@elmjag sure, we can have a quick chat if you like. How about tomorrow ? Otherwise I think the current proposal is fine maybe the color could be a bit more fun (other than dark gray :) ) and what do you think about a text perhaps under the hide button (saying "hidden by user" and perhaps "out of view" when hidden automatically) ?

mockoocy commented:
Undrafting.
As per @marcus-oscarsson's suggestion I changed the background color for table with grid info. 
Additionally I added a tooltip that shows the reason why a grid is hidden. I think it's nicer and informative enough :) 

![image](https://github.com/user-attachments/assets/c75fb8c6-e7b5-452e-868e-eb3073b678ae)
![withTooltip](https://github.com/user-attachments/assets/99ebb41f-62b2-4a29-8f9e-f13dfe0c6495)

marcus-oscarsson commented:
Would it be possible to get some more distinct color difference, between "hidden by user" and "out of range", like light blue and light green or something ?

marcus-oscarsson commented:
Or maybe no change for "our of range" and light blue or green for "hidden by user" ?

mockoocy commented:
> Or maybe no change for "our of range" and light blue or green for "hidden by user" ?

Light green's already reserved for selected grid, so I guess we'll go with light blue  :smile_cat:   
![image](https://github.com/user-attachments/assets/fdcb7957-c8e5-49eb-b568-fdef82defac5)

marcus-oscarsson commented:
Ok, yeah that looks better I guess, does it not ?. Would it make sense to make the text of the hidden meshes italic ?

mockoocy commented:
> Ok, yeah that looks better I guess, does it not ?. Would it make sense to make the text of the hidden meshes italic ?

Yeah, that's a good idea - done :)

marcus-oscarsson commented:
Cool, thanks alot very nice !

marcus-oscarsson commented on ui/src/components/SampleView/GridForm.jsx:
```diff
@@ -5,10 +5,27 @@ import Draggable from 'react-draggable';
+/**
+ * Returns the tooltip content for the hide button
+ *
+ * @param {boolean} userHidden - if the grid is hidden by the user
```
:+1: Nice, thanks

marcus-oscarsson commented on ui/src/components/SampleView/SampleView.css:
```diff
@@ -164,8 +164,13 @@
   font-weight: bold;
 }
+
+table.table-striped > tbody > tr[data-user-hidden='true'] {
```
:+1:

marcus-oscarsson commented:
I guess this is good to go then ?, Ill merge  https://github.com/mxcube/mxcubecore/pull/1044 then you can update `pyrproject.toml`

marcus-oscarsson commented:
My apologies it took sometime to get PR on `mxcubecore` merged, its done now :)
